File: examples/power/unmanaged_element/rust/src/unmanaged_element_integration_test.rs
// Copyright 2024 The Fuchsia Authors. Archivell rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use anyhow::{anyhow, Result};
use diagnostics_assertions::assert_data_tree;
use diagnostics_reader::{ArchiveReader, Inspect};
use fidl::endpoints::create_proxy;
use fuchsia_component_test::{Capability, ChildOptions, RealmBuilder, RealmInstance, Ref, Route};
use fuchsia_inspect::hierarchy::{DiagnosticsHierarchy, Property};
use futures::channel::mpsc;
use futures::StreamExt;
use unmanaged_element::UnmanagedElement;
use {
    fidl_fuchsia_power_broker as fbroker, fuchsia_async as fasync, power_broker_client as pbclient,
};

/// Manages a test Realm that includes a full Power Broker instance.
struct TestBrokerRealm {
    topology: fbroker::TopologyProxy,
    archive_reader: ArchiveReader,
    _realm: RealmInstance,
}

impl TestBrokerRealm {
    async fn build() -> Result<Self> {
        let builder = RealmBuilder::new().await?;

        // Add Power Broker component to this test realm.
        let power_broker_name = "test-power-broker";
        let power_broker = builder
            .add_child(power_broker_name, "#meta/power-broker.cm", ChildOptions::new())
            .await?;

        builder
            .add_route(
                Route::new()
                    .capability(Capability::protocol_by_name("fuchsia.power.broker.Topology"))
                    .from(&power_broker)
                    .to(Ref::parent()),
            )
            .await?;

        let _realm = builder.build().await?;
        let topology =
            _realm.root.connect_to_protocol_at_exposed_dir::<fbroker::TopologyMarker>()?;
        let mut archive_reader = ArchiveReader::new();
        archive_reader.add_selector(format!(
            "realm_builder\\:{}/{}:root",
            _realm.root.child_name(),
            power_broker_name
        ));
        Ok(Self { topology, archive_reader, _realm })
    }

    /// Should be called at the end of each test to ensure proper teardown between test cases.
    async fn destroy(self) -> Result<()> {
        Ok(self._realm.destroy().await?)
    }
}

/// Gets an element's current power level represented in a Power Broker Inspect Tree.
async fn check_inspect_element_current_level(
    broker_archive_reader: &ArchiveReader,
    element_name: &'static str,
) -> Result<fbroker::PowerLevel> {
    let root = broker_archive_reader
        .snapshot::<Inspect>()
        .await?
        .into_iter()
        .next()
        .and_then(|result| result.payload)
        .expect("snapshot of inspect hierarchy");
    let meta = find_element_meta_node(&root, element_name).expect("finding element node");
    if let Some(Property::Uint(_key, current_level)) = meta.get_property("current_level") {
        Ok(current_level.to_le_bytes()[0])
    } else {
        Err(anyhow!("failed to find current_level property"))
    }
}

/// Searches the provided Inspect hierarchy for a `meta` node matching:
/// `root/broker/topology/fuchsia.inspect.Graph/topology/<ID string>/meta`.
fn find_element_meta_node<'a>(
    root: &'a DiagnosticsHierarchy,
    name: &'a str,
) -> Option<&'a DiagnosticsHierarchy> {
    assert_data_tree!(root, root: contains {
        broker: contains {
            topology: contains { "fuchsia.inspect.Graph": contains { topology: contains {} } } } });
    let topology_node = root
        .get_child_by_path(&["broker", "topology", "fuchsia.inspect.Graph", "topology"])
        .expect("finding topology node");
    for child in topology_node.children.iter() {
        if let Some(meta) = child.get_child("meta") {
            if let Some(Property::String(_key, meta_name)) = meta.get_property("name") {
                if meta_name == name {
                    return Some(meta);
                }
            }
        }
    }
    None
}

#[fuchsia::test]
async fn unmanaged_element_state_propagates_to_status_channel() -> Result<()> {
    let broker_realm = TestBrokerRealm::build().await?;

    // Create the unmanaged power element.
    let element = UnmanagedElement::add(
        &broker_realm.topology,
        "unmanaged-element",
        &pbclient::BINARY_POWER_LEVELS,
        &pbclient::BINARY_POWER_LEVELS[0],
    )
    .await?;

    // Use the unmanaged element's ElementControl client to open a Status channel.
    let (status, status_server) = create_proxy::<fbroker::StatusMarker>();
    element.context.element_control.open_status_channel(status_server)?;

    // Listen for updates on the Status channel.
    let (status_sender, mut status_receiver) = mpsc::unbounded::<fbroker::PowerLevel>();
    let status_task = fasync::Task::spawn(async move {
        loop {
            let power_level = status
                .watch_power_level()
                .await
                .expect("fidl response")
                .expect("power level status");
            status_sender.unbounded_send(power_level).expect("sending status");
        }
    });
    assert_eq!(status_receiver.next().await, Some(pbclient::BINARY_POWER_LEVELS[0]));

    // Change power levels on the unmanaged element and observe updates on the Status channel.
    element.set_level(&pbclient::BINARY_POWER_LEVELS[1]).await?;
    assert_eq!(status_receiver.next().await, Some(pbclient::BINARY_POWER_LEVELS[1]));
    element.set_level(&pbclient::BINARY_POWER_LEVELS[0]).await?;
    assert_eq!(status_receiver.next().await, Some(pbclient::BINARY_POWER_LEVELS[0]));

    drop(status_task);
    broker_realm.destroy().await?;
    Ok(())
}

#[fuchsia::test]
async fn unmanaged_element_state_propagates_to_power_broker_inspect() -> Result<()> {
    let broker_realm = TestBrokerRealm::build().await?;

    // Create the unmanaged power element.
    let element_name = "unmanaged_element";
    let element = UnmanagedElement::add(
        &broker_realm.topology,
        element_name,
        &pbclient::BINARY_POWER_LEVELS,
        &pbclient::BINARY_POWER_LEVELS[0],
    )
    .await?;

    // Check initial power level in Power Broker's Inspect tree.
    assert_eq!(
        check_inspect_element_current_level(&broker_realm.archive_reader, element_name).await?,
        pbclient::BINARY_POWER_LEVELS[0]
    );

    // Change power levels on the unmanaged element and observe updates in Broker Inspect tree.
    element.set_level(&pbclient::BINARY_POWER_LEVELS[1]).await?;
    assert_eq!(
        check_inspect_element_current_level(&broker_realm.archive_reader, element_name).await?,
        pbclient::BINARY_POWER_LEVELS[1]
    );
    element.set_level(&pbclient::BINARY_POWER_LEVELS[0]).await?;
    assert_eq!(
        check_inspect_element_current_level(&broker_realm.archive_reader, element_name).await?,
        pbclient::BINARY_POWER_LEVELS[0]
    );

    broker_realm.destroy().await?;
    Ok(())
}

File: examples/power/unmanaged_element/cpp/unmanaged_element_integration_test.cc
// Copyright 2024 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <fidl/fuchsia.power.broker/cpp/fidl.h>
#include <lib/async/cpp/executor.h>
#include <lib/diagnostics/reader/cpp/archive_reader.h>
#include <lib/diagnostics/reader/cpp/inspect.h>
#include <lib/fidl/cpp/client.h>
#include <lib/inspect/cpp/hierarchy.h>
#include <lib/sys/component/cpp/testing/realm_builder.h>

#include <gtest/gtest.h>
#include <sdk/lib/driver/power/cpp/testing/scoped_background_loop.h>
#include <src/lib/testing/loop_fixture/real_loop_fixture.h>

#include "examples/power/unmanaged_element/cpp/unmanaged_element.h"

namespace {

using component_testing::ChildRef;
using component_testing::ParentRef;
using component_testing::Protocol;
using component_testing::RealmBuilder;
using component_testing::RealmRoot;
using component_testing::Route;
using diagnostics::reader::ArchiveReader;
using diagnostics::reader::InspectData;
using examples::power::UnmanagedElement;
using fdf_power::testing::ScopedBackgroundLoop;
using fuchsia_power_broker::BinaryPowerLevel;
using fuchsia_power_broker::PowerLevel;
using fuchsia_power_broker::Status;
using fuchsia_power_broker::Topology;
using inspect::IntPropertyValue;
using inspect::NodeValue;
using inspect::StringPropertyValue;

const std::vector<PowerLevel> kBinaryPowerLevels = {fidl::ToUnderlying(BinaryPowerLevel::kOff),
                                                    fidl::ToUnderlying(BinaryPowerLevel::kOn)};

class UnmanagedElementIntegrationTest : public gtest::RealLoopFixture {};

class TestBrokerRealm {
 public:
  explicit TestBrokerRealm(async_dispatcher_t* dispatcher)
      : realm_(RealmBuilder::Create()
                   .AddChild(power_broker_.name.data(), "#meta/power-broker.cm")
                   .AddRoute(Route{.capabilities = {Protocol{"fuchsia.power.broker.Topology"}},
                                   .source = power_broker_,
                                   .targets = {ParentRef()}})
                   .Build()) {}

  template <typename Protocol>
  fidl::ClientEnd<Protocol> TakeClient() {
    zx::result<fidl::ClientEnd<Protocol>> client = realm_.component().Connect<Protocol>();
    EXPECT_TRUE(client.is_ok()) << client.error_value();
    return std::move(client.value());
  }

  std::string GetInspectMoniker() const {
    return "realm_builder\\:" + realm_.component().GetChildName() + "/" + power_broker_.name.data();
  }

 private:
  ChildRef power_broker_{"test-power-broker"};
  RealmRoot realm_;
};

void CheckCurrentLevel(fidl::ClientEnd<Status>& status, const PowerLevel& expected_level) {
  auto result = fidl::WireCall(status)->WatchPowerLevel();
  EXPECT_TRUE(result->is_ok());
  ASSERT_EQ(result->value()->current_level, expected_level);
}

TEST_F(UnmanagedElementIntegrationTest, UnmanagedElementStatePropagatesToStatusChannel) {
  ScopedBackgroundLoop background;
  TestBrokerRealm broker_realm(background.dispatcher());
  const PowerLevel initial_level = kBinaryPowerLevels[0];
  auto topology = broker_realm.TakeClient<Topology>();
  auto element =
      UnmanagedElement::Add(topology, "unmanaged-element", kBinaryPowerLevels, initial_level);

  // Use the unmanaged element's ElementControl client to open a Status channel.
  auto [status_channel, status_channel_server] = fidl::Endpoints<Status>::Create();
  auto& element_control = element->description().element_control_client;
  EXPECT_TRUE(element_control.has_value());
  EXPECT_TRUE(
      fidl::WireCall(*element_control)->OpenStatusChannel(std::move(status_channel_server)).ok());

  // Listen for updates on the Status channel.
  CheckCurrentLevel(status_channel, initial_level);

  // Change power levels on the unmanaged element and observe updates on the Status channel.
  element->SetLevel(kBinaryPowerLevels[1]);
  CheckCurrentLevel(status_channel, kBinaryPowerLevels[1]);
  element->SetLevel(kBinaryPowerLevels[0]);
  CheckCurrentLevel(status_channel, kBinaryPowerLevels[0]);
}

/// Searches the provided Inspect hierarchy for a `meta` node matching:
/// `root/broker/topology/fuchsia.inspect.Graph/topology/<ID string>/meta`.
void CheckTopologyInspectNode(const fpromise::result<std::vector<InspectData>, std::string>& result,
                              const std::string& name, const PowerLevel& expected_level) {
  EXPECT_TRUE(result.is_ok()) << result.error();
  EXPECT_EQ(result.value().size(), 1u);
  const InspectData& data = result.value().front();
  EXPECT_TRUE(data.payload().has_value());
  const inspect::Hierarchy* topology_node = data.payload().value()->GetByPath(
      {"broker", "topology", "fuchsia.inspect.Graph", "topology"});
  EXPECT_NE(topology_node, nullptr);
  for (const inspect::Hierarchy& child : topology_node->children()) {
    const inspect::Hierarchy* meta = child.GetByPath({"meta"});
    if (meta) {
      const NodeValue& meta_node = meta->node();
      const StringPropertyValue* meta_name = meta_node.get_property<StringPropertyValue>("name");
      if (meta_name != nullptr && meta_name->value() == name) {
        const IntPropertyValue* current_level =
            meta_node.get_property<IntPropertyValue>("current_level");
        EXPECT_NE(current_level, nullptr);
        ASSERT_EQ(current_level->value(), expected_level);
        return;
      }
    }
  }
  FAIL() << "Power Broker Inspect tree missing element with current_level: " << name;
}

TEST_F(UnmanagedElementIntegrationTest, UnmanagedElementStatePropagatesToPowerBrokerInspect) {
  ScopedBackgroundLoop background;
  TestBrokerRealm broker_realm(background.dispatcher());
  const PowerLevel initial_level = kBinaryPowerLevels[0];
  auto topology = broker_realm.TakeClient<Topology>();
  const std::string element_name = "unmanaged-element";
  auto element = UnmanagedElement::Add(topology, element_name, kBinaryPowerLevels, initial_level);

  async::Executor executor(dispatcher());
  ArchiveReader reader(dispatcher(), {broker_realm.GetInspectMoniker() + ":root"});

  // Wait for Inspect data to be ready.
  bool ready = false;
  while (!ready) {
    executor.schedule_task(reader.GetInspectSnapshot().then(
        [&ready](fpromise::result<std::vector<InspectData>, std::string>& result) {
          EXPECT_TRUE(result.is_ok()) << result.error();
          ready = !result.value().empty();
        }));
    RunLoopWithTimeoutOrUntil([&ready] { return ready; });
  }

  // Check initial power level in Power Broker's Inspect tree.
  executor.schedule_task(reader.GetInspectSnapshot().then(
      [&element_name,
       quit = QuitLoopClosure()](fpromise::result<std::vector<InspectData>, std::string>& result) {
        CheckTopologyInspectNode(result, element_name, kBinaryPowerLevels[0]);
        quit();
      }));
  RunLoop();

  // Change power levels on the unmanaged element and observe updates in Broker Inspect
  // tree.
  element->SetLevel(kBinaryPowerLevels[1]);
  executor.schedule_task(reader.GetInspectSnapshot().then(
      [&element_name,
       quit = QuitLoopClosure()](fpromise::result<std::vector<InspectData>, std::string>& result) {
        CheckTopologyInspectNode(result, element_name, kBinaryPowerLevels[1]);
        quit();
      }));
  RunLoop();
  element->SetLevel(kBinaryPowerLevels[0]);
  executor.schedule_task(reader.GetInspectSnapshot().then(
      [&element_name,
       quit = QuitLoopClosure()](fpromise::result<std::vector<InspectData>, std::string>& result) {
        CheckTopologyInspectNode(result, element_name, kBinaryPowerLevels[0]);
        quit();
      }));
  RunLoop();
}

}  // namespace
